File: javascript_files/raylib/pong.js
const r = require("raylib");

const screenWidth = 800;
const screenHeight = 600;
const screenTitle = "Pong";
const screenBackground = r.SKYBLUE;
const gameFps = 60;

/*
    PADDLE
*/
class Paddle {
  constructor(x) {
    this.width = 10;
    this.height = 100;
    this.initial_x = x;
    this.initial_y = r.GetScreenHeight() / 2 - this.height / 2;
    this.x = this.initial_x;
    this.y = this.initial_y;
    this.color = r.BLACK;
    this.speed = 5;
    this.ai_speed = 7;
  }

  reset() {
    this.x = this.initial_x;
    this.y = this.initial_y;
  }

  draw() {
    r.DrawRectangle(this.x, this.y, this.width, this.height, this.color);
  }

  update() {
    if (r.IsKeyDown(r.KEY_W) && this.y >= 0) {
      this.y -= this.speed;
    } else if (
      r.IsKeyDown(r.KEY_S) &&
      this.y <= r.GetScreenHeight() - this.height
    ) {
      this.y += this.speed;
    }
  }

  ai_update(ball) {
    if (ball.x > r.GetScreenWidth() / 2) {
      let distance_x = ball.x - (this.x + this.width / 2);
      let distance_y = ball.y - (this.y + this.height / 2);
      let distance = Math.sqrt(distance_x ** 2 + distance_y ** 2);

      if (distance > 0) {
        distance_y /= distance;

        this.y += distance_y * this.ai_speed;
      }
    }
  }
}

/*
    BALL
*/
class ball {
  constructor() {
    this.initial_x = r.GetScreenWidth() / 2;
    this.initial_y = r.GetScreenHeight() / 2;
    this.x = this.initial_x;
    this.y = this.initial_y;
    this.radius = 10;
    this.color = r.RED;
    this.initial_speed_x = 5;
    this.initial_speed_y = 6;
    this.speed_x = this.initial_speed_x;
    this.speed_y = this.initial_speed_y;
    this.speed_interval = 0.005;
    this.frames_counter = 0;
  }

  reset() {
    this.x = this.initial_x;
    this.y = this.initial_y;
    this.speed_x = this.initial_speed_x;
    this.speed_y = this.initial_speed_y;
    this.speed_x *= -1;
    this.frames_counter = 0;
  }

  draw() {
    r.DrawCircleV(r.Vector2(this.x, this.y), this.radius, this.color);
  }

  update() {
    this.frames_counter++;
    if (this.frames_counter >= 60) {
      this.frames_counter = 61;

      this.x += this.speed_x;
      this.y += this.speed_y;

      this.speed_x += this.speed_interval;
      this.speed_y += this.speed_interval;
    }

    if (this.y <= this.radius || this.y >= r.GetScreenHeight() - this.radius) {
      this.speed_y *= -1;
    }
  }
}

/*
    GAME
*/
class Game {
  constructor() {
    this.score_left = 0;
    this.score_right = 0;

    this.paddle_left = new Paddle(10);
    this.ai = new Paddle(r.GetScreenWidth() - this.paddle_left.width - 10);
    this.ball = new ball();
  }

  reset() {
    this.paddle_left.reset();
    this.ai.reset();
    this.ball.reset();
  }

  draw() {
    // draw scores
    r.DrawText(String(this.score_left), 200, 20, 40, r.BLACK);
    r.DrawText(
      String(this.score_right),
      r.GetScreenWidth() - 220,
      20,
      40,
      r.BLACK
    );

    this.paddle_left.draw();
    this.ai.draw();
    this.ball.draw();
  }

  update() {
    this.paddle_left.update();
    this.ai.ai_update(this.ball);
    this.ball.update();

    // ball collision paddle
    if (
      r.CheckCollisionCircleRec(
        r.Vector2(this.ball.x, this.ball.y),
        this.ball.radius,
        r.Rectangle(
          this.paddle_left.x,
          this.paddle_left.y,
          this.paddle_left.width,
          this.paddle_left.height
        )
      )
    ) {
      this.ball.speed_x *= -1;
    } else if (
      r.CheckCollisionCircleRec(
        r.Vector2(this.ball.x, this.ball.y),
        this.ball.radius,
        r.Rectangle(this.ai.x, this.ai.y, this.ai.width, this.ai.height)
      )
    ) {
      this.ball.speed_x *= -1;
    }

    // update score
    if (this.ball.x < 0) {
      this.score_right++;
      this.reset();
    } else if (this.ball.x > r.GetScreenWidth()) {
      this.score_left++;
      this.reset();
    }
  }
}

/*
    RAYLIB WINDOW
*/
function init() {
  r.InitWindow(screenWidth, screenHeight, screenTitle);
  r.SetTargetFPS(gameFps);

  const game = new Game();

  while (!r.WindowShouldClose()) {
    r.BeginDrawing();
    r.ClearBackground(screenBackground);

    game.draw();
    game.update();

    r.EndDrawing();
  }

  r.CloseWindow();
}

init();

File: c_files/raylib/single_file_games/pong.c
#include "raylib.h"
#include <math.h>
#include <stdio.h>

#define SCORE_STR_LENGTH 10

//**************************************
// BALL
//**************************************

typedef struct Ball
{
    float initial_x;
    float initial_y;
    float x;
    float y;
    float radius;
    Color color;
    float initial_speed_x;
    float initial_speed_y;
    float speed_x;
    float speed_y;
    float speed_increment;
    bool active;

} Ball;

void init_ball(Ball *ball)
{
    ball->initial_x = (float)GetScreenWidth() / 2;
    ball->initial_y = (float)GetScreenHeight() / 2;
    ball->x = ball->initial_x;
    ball->y = ball->initial_y;
    ball->radius = 10;
    ball->color = RED;
    ball->initial_speed_x = 4.0f;
    ball->initial_speed_y = 4.0f;
    ball->speed_x = ball->initial_speed_x;
    ball->speed_y = ball->initial_speed_y;
    ball->speed_increment = 0.0005f;
    ball->active = false;
}

void reset_ball(Ball *ball)
{
    ball->x = ball->initial_x;
    ball->y = ball->initial_y;
    ball->speed_x = ball->initial_speed_x;
    ball->speed_y = ball->initial_speed_y;
    ball->speed_x *= -1;
    ball->active = false;
}

void draw_ball(Ball *ball)
{
    DrawCircleV((Vector2){ball->x, ball->y}, ball->radius, ball->color);
}

void update_ball(Ball *ball)
{
    if (IsKeyPressed(KEY_SPACE))
    {
        ball->active = true;
    }

    if (ball->active)
    {
        ball->x += ball->speed_x;
        ball->y += ball->speed_y;

        // increment ball speed
        if (ball->speed_x > 0)
        {
            ball->speed_x += ball->speed_increment;
        }
        else if (ball->speed_x < 0)
        {
            ball->speed_x -= ball->speed_increment;
        }
        else if (ball->speed_y > 0)
        {
            ball->speed_x += ball->speed_increment;
        }
        else if (ball->speed_x < 0)
        {
            ball->speed_x -= ball->speed_increment;
        }
    }

    // ball vertical bounds
    if (ball->y <= ball->radius || ball->y >= GetScreenHeight() - ball->radius)
    {
        ball->speed_y *= -1;
    }
}

//**************************************
// PADDLE
//**************************************

typedef struct Paddle
{
    float initial_x;
    float initial_y;
    float x;
    float y;
    float width;
    float height;
    Color color;
    float speed;
    float ai_speed;

} Paddle;

void init_paddle(Paddle *paddle, float x)
{
    paddle->width = 10.0f;
    paddle->height = 100.0f;
    paddle->initial_x = x;
    paddle->initial_y = (float)GetScreenHeight() / 2 - paddle->height / 2;
    paddle->x = paddle->initial_x;
    paddle->y = paddle->initial_y;
    paddle->color = BLACK;
    paddle->speed = 5.0f;
    paddle->ai_speed = 7.0f;
}

void reset_paddle(Paddle *paddle)
{
    paddle->x = paddle->initial_x;
    paddle->y = paddle->initial_y;
}

void draw_paddle(Paddle *paddle)
{
    DrawRectangleRec((Rectangle){paddle->x,
                                 paddle->y,
                                 paddle->width,
                                 paddle->height},
                     paddle->color);
}

void update_paddle_left(Paddle *paddle_left)
{
    if (IsKeyDown(KEY_W) && paddle_left->y >= 0)
    {
        paddle_left->y -= paddle_left->speed;
    }
    else if (IsKeyDown(KEY_S) && paddle_left->y <= GetScreenHeight() - paddle_left->height)
    {
        paddle_left->y += paddle_left->speed;
    }
}

void update_paddle_ai(Paddle *paddle_ai, Ball *ball)
{
    if (ball->x > GetScreenWidth() / 2)
    {
        float distance_x = ball->x - (paddle_ai->x + paddle_ai->width / 2);
        float distance_y = ball->y - (paddle_ai->y + paddle_ai->height / 2);
        float distance = sqrt(pow(distance_x, 2) + pow(distance_y, 2));
        if (distance > 0)
        {
            distance_y /= distance;

            paddle_ai->y += distance_y * paddle_ai->ai_speed;
        }
    }
}

//**************************************
// MAIN LOOP
//**************************************

int main(void)
{
    const int screenWidth = 800;
    const int screenHeight = 600;
    const char screenTitle[] = "Pong";
    const Color screenBackground = SKYBLUE;
    const int gameFps = 60;

    InitWindow(screenWidth, screenHeight, screenTitle);
    SetTargetFPS(gameFps);

    int score_left = 0;
    int score_right = 0;
    char score_left_str[SCORE_STR_LENGTH];
    char score_right_str[SCORE_STR_LENGTH];

    Paddle paddle_left;
    Paddle paddle_ai;
    Ball ball;

    init_paddle(&paddle_left, 10.0f);
    init_paddle(&paddle_ai, GetScreenWidth() - paddle_left.width - 10.0f);
    init_ball(&ball);

    while (!WindowShouldClose())
    {
        // convert score to string
        snprintf(score_left_str, SCORE_STR_LENGTH, "%d\n", score_left);
        snprintf(score_right_str, SCORE_STR_LENGTH, "%d\n", score_right);

        update_paddle_left(&paddle_left);
        update_paddle_ai(&paddle_ai, &ball);
        update_ball(&ball);

        // ball horizontal bounds
        if (ball.x < 0)
        {
            score_right++;
            reset_paddle(&paddle_left);
            reset_paddle(&paddle_ai);
            reset_ball(&ball);
        }
        else if (ball.x > GetScreenWidth())
        {
            score_left++;
            reset_paddle(&paddle_left);
            reset_paddle(&paddle_ai);
            reset_ball(&ball);
        }

        // ball collision paddles
        if (CheckCollisionCircleRec((Vector2){ball.x, ball.y},
                                    ball.radius,
                                    (Rectangle){paddle_left.x,
                                                paddle_left.y,
                                                paddle_left.width,
                                                paddle_left.height}))
        {
            ball.speed_x *= -1;
        }
        else if (CheckCollisionCircleRec((Vector2){ball.x, ball.y},
                                         ball.radius,
                                         (Rectangle){paddle_ai.x,
                                                     paddle_ai.y,
                                                     paddle_ai.width,
                                                     paddle_ai.height}))
        {
            ball.speed_x *= -1;
        }

        BeginDrawing();
        ClearBackground(screenBackground);

        // draw scores
        DrawText(score_left_str, 200, 30, 40, BLACK);
        DrawText(score_right_str, GetScreenWidth() - 220, 30, 40, BLACK);

        draw_paddle(&paddle_left);
        draw_paddle(&paddle_ai);
        draw_ball(&ball);

        EndDrawing();
    }

    CloseWindow();

    return 0;
}